zerver/actions/pinned_views.py

```diff
@@ -0,0 +1,83 @@
+from typing import Any
+
+from django.db import transaction
+from django.utils.translation import gettext as _
+
+from zerver.lib.exceptions import JsonableError, ResourceNotFoundError
+from zerver.models import PinnedView, UserProfile
+from zerver.models.pinned_views import LeftSidebarViewLocationEnum
+from zerver.tornado.django_api import send_event_on_commit
+
+VALID_VIEW_IDS = ["inbox", "recent", "reactions", "mentions", "feed", "drafts", "starred"]
+
+
+def validate_sidebar_location(location: int) -> None:
+    if location not in LeftSidebarViewLocationEnum.__members__.values():
+        raise JsonableError(_("Invalid location."))
+
+
+def validate_sidebar_view_id(view_id: str) -> None:
+    if view_id not in VALID_VIEW_IDS:
+        raise JsonableError(_("Invalid view_id."))
+
+
+@transaction.atomic(durable=True)
+def do_add_pinned_view(
+    user_profile: UserProfile,
+    view_id: str,
+    location: int,
+    name: str | None = None,
+    url_hash: str | None = None,
+) -> PinnedView:
+    validate_sidebar_location(location)
+    validate_sidebar_view_id(view_id)
```

Comment:
Generally, we do this sort of validation in the view code with Pydantic, and pass the resolved value here. `do_` methods are not intended to fail because of input validation issues.